### PR TITLE
feat: improve load_requirements in setup.py

### PR DESCRIPTION
**Pull request by rbiseck3**
### Description
This PR adds a number of enhancements around how we dynamically generate the requirements defined in the `setup.py` file. All logic around this was moved out into a `setup_utils.py` file to support unit tests around the logic:
 
* `load_requirements`: Current approach to dynamically generate the requirements was causing them to maintain the newline character at the end. This was removed. Also added support for the `-r` character in `*.in` files which recursively pull in other `.in` files which would currently be ignored. This isn't being used at the moment so it wasn't causing a problem, but this is future looking. It also used to support a list of filenames but that was never used, isolated to take in only a string filename of a Path object.
* New methods added to generate the dictionaries fo requirements, one for connector requirements and another for doc type requirements. 
* Safegaurds added throughout to raise an error if something incorrect was added. These include the file not being found or it not having the expected `.in` suffix.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -32,11 +33,19 @@ def load_requirements(file_list: Optional[Union[str, List[str]]] = None) -> List
...
+            raw = f.read().splitlines()
+            requirements.extend([r for r in raw if not r.startswith("#") and not r.startswith("-")])
+            recursive_reqs = [r for r in raw if r.startswith("-r")]
+            if recursive_reqs:
+                filenames = []
+                for recursive_req in recursive_reqs:
+                    file_spec = recursive_req.split()[-1]
+                    file_path = Path(file_dir) / file_spec
+                    filenames.append(str(file_path.resolve()))
+                requirements.extend(load_requirements(file_list=filenames))
```
Comment (qued) [inline setup.py:48]: We don't need to keep the file open after we've read it, so this block can be moved out of the `with` clause.

File: setup.py [hunk 2]
```diff
@@ -32,11 +33,19 @@ def load_requirements(file_list: Optional[Union[str, List[str]]] = None) -> List
...
+        file_dir = path.parent.resolve()
         with open(file, encoding="utf-8") as f:
-            requirements.extend(f.readlines())
-    requirements = [
-        req for req in requirements if not req.startswith("#") and not req.startswith("-")
-    ]
+            raw = f.read().splitlines()
+            requirements.extend([r for r in raw if not r.startswith("#") and not r.startswith("-")])
+            recursive_reqs = [r for r in raw if r.startswith("-r")]
+            if recursive_reqs:
```
Comment (qued) [inline setup.py]: Small nit: the code iterates over `recursive_reqs`, so if `recursive_reqs` is empty all the code within this `if` block is a noop anyway. In such cases I think it's cleaner and more readable to not include the `if` statement.

Comment (rbiseck3) [reply setup.py]: I actually did this for readability because I only instantiate the `filenames = []` if there are any recursive ones to look through.

Comment (rbiseck3) [reply setup.py]: But I could just have it call it one at a time.

File: test_unstructured/test_setup_utils.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from setup_utils import load_requirements
+
+
+def test_load_requirements():
+    file = "./files/example.in"
```
Comment (qued) [inline test_unstructured/test_setup_utils.py:5]:
```suggestion
    file = f"{os.path.dirname(__file__)}/files/example.in"
```
When dealing with tests, the working directory gets wonky. In many situations, it's the project base directory. So this test fails when, for example, I run `pytest test_unstructured/test_setup.py`. Maybe it passes when I `make test`... but I don't always want to run the whole suite. If you follow this suggestion you'd also need to `import os`.

File: setup_utils.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            raw = f.read().splitlines()
+            requirements.extend([r for r in raw if not r.startswith("#") and not r.startswith("-")])
+            recursive_reqs = [r for r in raw if r.startswith("-r")]
+            if recursive_reqs:
+                filenames = []
+                for recursive_req in recursive_reqs:
+                    file_spec = recursive_req.split()[-1]
+                    file_path = Path(file_dir) / file_spec
+                    filenames.append(str(file_path.resolve()))
+                requirements.extend(load_requirements(file_list=filenames))
```
Comment (qued) [inline setup_utils.py]: Just reiterating we don't want this section in the `with` clause since that keeps the file open unnecessarily.

Comment (rbiseck3) [reply setup_utils.py]: Yep, this was moved out

**Comment by rbiseck3** [comment]
@qued, repo uploaded to test pypi here:  [unstructured/0.13.3.dev3/](https://test.pypi.org/project/unstructured/0.13.3.dev3/)

**Pull request closed without merging**